### PR TITLE
[ZEPPELIN-1629] Enable renaming folder from the main page

### PR DESCRIPTION
**Pull request by tae-jun**
### What is this PR for?
This PR will enable renaming folder from the main page.

I wrote some codes. Hope this PR is the start of folder-based features. (e.g folder authorization, folder focused view)

Please check when you have some time!
Thanks 😄

### What type of PR is it?
[Feature]

### Todos
* [x] - Fix ConcurrentModificationException of FolderView

### What is the Jira issue?
[ZEPPELIN-1629](https://issues.apache.org/jira/browse/ZEPPELIN-1629)

### How should this be tested?
Change folder names!

### Screenshots (if appropriate)
rename_folder

### Questions:
* Does the licenses files need update? no
* Is there breaking changes for older versions? no
* Does this needs documentation? no

### CODE REVIEW

**Comment by tae-jun** [comment]
Fixed error and CI is green!

Ready for review! 😄

**Comment by AhyoungRyu** [comment]
Awesome. Let me test this. Will ping you again :)

**Comment by AhyoungRyu** [comment]
@tae-jun Sorry for my late response.
I tested but can't edit the dir name in root dir level. The folder name can be modified only in below level. Please see the below attached gif img.
dirname

**Comment by tae-jun** [comment]
@AhyoungRyu Thanks for review 😄 Don't need to hurry! I will check and ping you :)

**Comment by tae-jun** [comment]
@AhyoungRyu ping ping~ I tested!

rename_folder_for_rootdir

But it just works fine with me. I tested both running ZeppelinServer from IntelliJ and binary built from `mvn clean package -DskipTests`

I use macOS 10.12.1 and Chrome. Could you please share your test environment? 😄

**Comment by AhyoungRyu** [comment]
@tae-jun Build it again and tested, but It works well as you intended. It was my bad. 
I tested two cases: \w & \wt Shiro. Both are works well.

Probably it's a nitpick,
screen shot 2016-11-18 at 7 00 52 pm

when I try to rename the folder name at home screen, the above msg is shown up. Could you change `renameFolder Notebook` to `rename folder`? (I know `renameFolder` comes from `op` and `notebook` is hardcoded, but sounds a bit awkward ㅜ_ㅜ)

**Comment by tae-jun** [comment]
@AhyoungRyu Thanks for your kind review. I appreciate it.

Actually, I didn't test with Shiro ^^; and didn't know what `op` parameter means. So thanks very much.

I changed `renameFolder` to `rename folder of` you addressed and also `renameNote` to `rename`.

Now it looks like this:




I also included note name which needs permission 😄

What do you think?

**Comment by AhyoungRyu** [comment]
@tae-jun It works well when I tried to rename folder name. And I think It's good idea to let user know which notes that he don't have permission exactly. Thanks :)
But currently CI is failed with below compilation error. Could you take a look again?
```
[ERROR] COMPILATION ERROR : 
[INFO] -------------------------------------------------------------
[ERROR] /home/travis/build/apache/zeppelin/zeppelin-server/src/main/java/org/apache/zeppelin/socket/NotebookServer.java:[774,5] cannot find symbol
  symbol:   class Folder
  location: class org.apache.zeppelin.socket.NotebookServer
[INFO] 1 error
```

**Comment by AhyoungRyu** [comment]
@tae-jun And one more thing, it would be better to keep consistency between `Note` and `Notebook`(combination of several notes). I pushed one commit to your branch, could you please take a look https://github.com/tae-jun/zeppelin/pull/1?

It will be shown when user rename the folder(I removed `notebook`):
screen shot 2016-11-20 at 10 25 22 am

And then user rename the note:
screen shot 2016-11-20 at 10 25 05 am

And also when user create a note, I changed `Example: NoteDir1/Notebook1` -> `NoteDir1/Note1`
screen shot 2016-11-20 at 10 24 01 am

**Comment by tae-jun** [comment]
@AhyoungRyu Thanks! I also think `notebook` is a little bit weird :) I merged your PR thanks.

This change will affect other insufficient privileges error messages as well:
* Insufficient privileges to `read` ~~notebook~~
* Insufficient privileges to `update` ~~notebook~~
* Insufficient privileges to `remove` ~~notebook~~
* Insufficient privileges to `write` ~~notebook~~
* Insufficient privileges to `clear output` ~~notebook~~

I prefer removing `notebook`. What do you think guys? If anyone has a concern about this, please tell.

And about CI, I suffered from that error when I tried to compile only one project (e.g. `zeppelin-server`). Since `zeppelin-server` depends on `zeppelin-interpreter` and `zeppelin-zengine`, I had to build two dependencies and `install` them on local maven repo, which can be done with `mvn clean package install -DskipTests`. If not, when it tries to compile `zeppelin-server`, it cannot find symbols of `zeppelin-zengine` and goes to fail.

However, I'm not sure that CI only recompiles changed files. I will look into it, and tell you if I get some more information.
(CI was green before 😭)

File: zeppelin-server/src/main/java/org/apache/zeppelin/socket/NotebookServer.java [hunk 1]
```diff
@@ -713,7 +713,7 @@ private void renameNote(NotebookSocket conn, HashSet<String> userAndRoles,
 
     NotebookAuthorization notebookAuthorization = notebook.getNotebookAuthorization();
     if (!notebookAuthorization.isOwner(noteId, userAndRoles)) {
-      permissionError(conn, "rename", fromMessage.principal,
+      permissionError(conn, "rename the note", fromMessage.principal,
```
Comment (AhyoungRyu) [inline zeppelin-server/src/main/java/org/apache/zeppelin/socket/NotebookServer.java]: @tae-jun I didn't care about the other operations [as you said](https://github.com/apache/zeppelin/pull/1630#issuecomment-261773204). It was my bad. I just thought about the renaming.. How about just let this "rename the note" -> "rename" as before, and put `note` in [here](https://github.com/apache/zeppelin/pull/1630/commits/12b0bae387dca3f3898a274a315312441f682e95#diff-bcbdda87aa29314ed459e8212067e313R605)?

**Comment by tae-jun** [comment]
@AhyoungRyu I fixed what you mentioned 😄 Please review.

**Comment by tae-jun** [comment]
I guess CI fails because it only recompiles changed files. (since build time is too short)

image

It should be about 30min.

Is there any way to compile the whole project again? I ran CI on [my own repository](https://travis-ci.org/tae-jun/zeppelin). It failed on some profiles because of Ignite interpreter. (what is relevance between Ignite interpreter and my PR? 😭) But it passed on other profiles whereas every profile failed on this repository.

**Comment by tae-jun** [comment]
@AhyoungRyu Yeah!!! Finally!!!

CI is green!!! It was my fault! I didn't import `Folder` at `zeppelin-server`.

(Why was it green before...?)

Please take a look!

File: zeppelin-server/src/main/java/org/apache/zeppelin/socket/NotebookServer.java [hunk 2]
```diff
@@ -48,13 +48,7 @@
...
 import org.apache.zeppelin.interpreter.remote.RemoteInterpreterProcessListener;
 import org.apache.zeppelin.interpreter.thrift.InterpreterCompletion;
-import org.apache.zeppelin.notebook.JobListenerFactory;
-import org.apache.zeppelin.notebook.Note;
-import org.apache.zeppelin.notebook.Notebook;
-import org.apache.zeppelin.notebook.NotebookAuthorization;
-import org.apache.zeppelin.notebook.NotebookEventListener;
-import org.apache.zeppelin.notebook.Paragraph;
-import org.apache.zeppelin.notebook.ParagraphJobListener;
+import org.apache.zeppelin.notebook.*;
```
Comment (AhyoungRyu) [inline zeppelin-server/src/main/java/org/apache/zeppelin/socket/NotebookServer.java]: Actually Zeppelin doesn't recommend using `*`. So if you missed `import org.apache.zeppelin.notebook.Folder` before,  will be better to add it only I think :)

**Comment by tae-jun** [comment]
@AhyoungRyu Fixed what you mentioned :) Always thanks for the review 😄

**Comment by AhyoungRyu** [comment]
@tae-jun Thanks for your prompt update. Appreciated your continues work. Tested again and checked the root folder naming issue is gone now. But when I create **over 3 depth** folder structure, then the root folder name can't be modified again. Here are the reproducing steps.
1. Create note with `/a/b/c`
2. Try to edit root folder `a` with `d` -> click `Rename` btn
3. It's not changed to `d`. Still remained as `a`

Could you check again? We're almost done I think!

**Comment by tae-jun** [comment]
@AhyoungRyu Sorry for my late response! But it's due to tests :)

I reproduced the error you mentioned and you were right!

Furthermore, chasing a reason of the error, I notice that there are more possibilities of error. So I wrote more codes and added more test cases.

## Folder merging test
rename_folder_merging

## Renaming a high depth folder test
rename_folder_complex2

I thought this new feature is simple to implement... but now I don't think so...^^; I tested a lot of times, but maybe there are more bugs. So please test this when you have some time.

Thanks 😄

**Comment by tae-jun** [comment]
CI failure is because of me. I will fix it :)

**Comment by AhyoungRyu** [comment]
@tae-jun Great work! Take your time and feel free to ping me again when you're ready.

**Comment by tae-jun** [comment]
@AhyoungRyu Yeah! CI is green 😄  Please review! Thanks

**Comment by soralee** [comment]
@tae-jun This is Cool! and It works well!
So, I suggest one thing. I think it would be nice to appear popup which is whether user will merge the folders or not before merging folders. In my case, It is familiar to me :) 

What do you think?

**Comment by tae-jun** [comment]
@soralee Thanks for the review! 😄 That's a great idea! I will give some effect on the front-end and ping you again. Thanks :)

**Comment by tae-jun** [comment]
@soralee Hi! I added the feature you suggested. What do you think? Is title or content of warning message appropriate? Thanks for the cool suggestion!

rename_folder_merging_warning

**Comment by AhyoungRyu** [comment]
@soralee What a nice suggestion! @tae-jun I think it's cool!!

**Comment by tae-jun** [comment]
@AhyoungRyu Always thanks for the review 😄

**Comment by soralee** [comment]
@tae-jun It is very AWESOME 👍 and thanks you for the accept my suggestion. 
Let me check again after CI is green 😄

**Comment by tae-jun** [comment]
@AhyoungRyu @soralee Yeah! CI is green!

**Comment by Leemoonsoo** [comment]
Tested and working really nicely. Looks awesome to me!
Thanks @tae-jun for the great contribution!

**Comment by tae-jun** [comment]
Thanks @Leemoonsoo! I appreciate your review 😄

**Comment by tae-jun** [comment]
I've tested more and it was fine most times. But occasionally it doesn't work well. I think it should be tested more! I don't know why for now but I will figure it out and ping when it's ready :)

**Comment by tae-jun** [comment]
I found out bugs by adding loggers and fixed them!

CI is green and it seems to work perfectly. I think it's good to go now 😄 

Please when you have some time. Thanks!

**Comment by Leemoonsoo** [comment]
Great work @tae-jun. LGTM!

**Comment by Leemoonsoo** [comment]
Merge to master if there're no more comments

**Comment by AhyoungRyu** [comment]
Tested and it's working nicely! LGTM 👍

**Pull request closed without merging**